File: 算法/查找/二分查找.py
def binary_search(arr: list, target: int):
    low = 0
    high = len(arr) - 1

    while low <= high:
        # 中间元素的位置
        mid = (low + high) // 2

        # 检查中间元素是否是目标元素
        if arr[mid] == target:
            return mid

        # 如果中间元素大于目标元素，缩小查找范围至左半部分
        elif arr[mid] > target:
            high = mid - 1

        # 如果中间元素小于目标元素，缩小查找范围至右半部分
        else:
            low = mid + 1

    # 目标元素不存在于数组中
    return None


if __name__ == '__main__':
    arr = [2, 5, 8, 12, 16, 23, 38, 56, 72, 91]
    target = 5

    result = binary_search(arr, target)

    if result == None:
        print("目标元素不存在于数组中")
    else:
        print("目标元素位于索引", result)

File: 算法/查找/二分查找.cpp
//
// Created by Momo on 2023/6/28.
//
#include <iostream>

using namespace std;

int binary_searh(const int * arr, int target, int len) {
    int low = 0;
    int high = len - 1;

    while (low <= high) {
        int mid = (low + high) / 2;
        // 中间位置就是我们要找的时候
        if (arr[mid] == target)
            return mid;
        // 目标值 < 中间值
        else if (arr[mid] > target)
            high = mid - 1;     // 范围压缩到左边的位置
        // 目标值 > 中间值
        else
            low = mid + 1;      // 范围压缩到右边的位置
    }
    return 0;
}


int main() {
    int arr[10] = {2, 5, 8, 12, 16, 23, 38, 56, 72, 91};

    int target = 5;

    int index = binary_searh(arr, target, sizeof(arr) / sizeof(arr[0]));

    if (index != false)
        cout << "Index is " << index
        << ", No. " << index + 1 << " in array"
        << endl;
    else
        cout<< "Target isn't in array" << endl;

    return 0;
}